### PR TITLE
yosys: add option to build with Python binding

### PR DESCRIPTION
**Pull request by povik**
## Description of changes

Yosys comes with Python bindings and the support to load Python plugins, though this is not enabled on most builds out in the world. Add this as an optional feature to the nixpkgs derivation. Hook up the derivation with the bindings enabled under ~~`yosysWithPython`~~ `yosys` and also as a Python module `python3Packages.yosys`. The first will be useful if people run `yosys`, then load a Python plugin. The second is useful if someone wants to use Yosys as a library from a Python program.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/compilers/yosys/default.nix [hunk 1]
```diff
@@ -18,6 +19,7 @@
 , yosys-bluespec
 , yosys-ghdl
 , yosys-symbiflow
+, enablePython ? false # enable python binding
```
Comment (thoughtpolice) [inline pkgs/development/compilers/yosys/default.nix:22]: Users of tools like `sby` will already need Python anyway; I think it's also enabled in the `nextpnr` build. We can probably go ahead and enable this safely, I think?

**Comment by thoughtpolice** [comment]
Thanks, see my comment. In general I have no issue with this.

File: pkgs/development/compilers/yosys/default.nix (hunk 1 above)
Comment (povik) [reply pkgs/development/compilers/yosys/default.nix:22]: I don’t think SBY (or any YosysHQ project) relies on the Python binding; it is written in Python but connects to Yosys in other ways. In any case I am not opposed to enabling the binding by default, the heavy boost dependency is the only downside I think.

**Comment by povik** [comment]
@thoughtpolice Thanks for taking a look. I have made the change to make building with the binding the default now.

File: pkgs/development/compilers/yosys/default.nix (hunk 1 above)
Comment (thoughtpolice) [reply pkgs/development/compilers/yosys/default.nix]: I know it doesn't rely on the python binding, I just meant to say "most users will practically have python in their closure already" for reasons like that, which means this is a very small thing to add, relatively.

**Review by thoughtpolice — APPROVED** [review]

File: pkgs/development/compilers/yosys/default.nix (hunk 1 above)
Comment (povik) [reply pkgs/development/compilers/yosys/default.nix]: Ah, yeah, didn't occur to me to think of users for who Python isn't pulled in one way or other already

**Comment by thoughtpolice** [comment]
Yeah, I think doing a passthru so that plugins get the right set of linkable dependencies is the right way forward. Would you be willing to write a patch for this? If not we may want to revert for a day or two until one of us can get some time to do it.

/cc @hzeller, who maintains the synlig work

**Comment by povik** [comment]
On second thought, instead of introducing a new passthru, isn't this exactly the job for `propagatedBuildInputs`? Whatever we decide, I should be able to write the patch for it.

**Pull request merged**